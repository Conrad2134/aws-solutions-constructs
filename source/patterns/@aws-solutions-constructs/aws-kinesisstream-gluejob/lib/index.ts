/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnJob, CfnJobProps } from '@aws-cdk/aws-glue';
import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Stream, StreamProps } from '@aws-cdk/aws-kinesis';
import { Bucket, BucketProps, CfnBucket } from '@aws-cdk/aws-s3';
import { Construct } from '@aws-cdk/core';
import * as defaults from '@aws-solutions-constructs/core';


export interface KinesisStreamGlueJobProps {
    /**
     * Existing instance of Kineses Data Stream. If not set, it will create an instance
     */
    readonly existingStreamObj?: Stream;
    /**
     * User provided props to override the default props for the Kinesis Stream.
     *
     * @default - Default props are used
     */
    readonly kinesisStreamProps?: StreamProps | any;
    /**
     * User provides props to override the default props for Glue ETL Jobs. This parameter will be ignored if the
     * existingGlueJob parameter is set
     *
     * @default - Default props are used
     */
    readonly glueJobProps?: CfnJobProps;
    /**
     * Existing GlueJob configuration. If not set, it will create the a CfnJob instance using the glueJobProps
     */
    readonly existingGlueJob?: CfnJob;
}

export class KinesisStreamGlueJob extends Construct {
    public readonly kinesisStream: Stream;

    public readonly glueJob: CfnJob;

    constructor(scope: Construct, id: string, props: KinesisStreamGlueJobProps) {
        super(scope, id);

        this.kinesisStream = defaults.buildKinesisStream(this, {
            existingStreamObj: props.existingStreamObj,
            kinesisStreamProps: props.kinesisStreamProps,
        });

        this.glueJob = defaults.buildGlueJob(this, {
            existingCfnJob: props.existingGlueJob,
            glueJobProps: props.glueJobProps
        });

        let _glueJobRoleName: string;
        if (props.existingGlueJob === undefined) {
            _glueJobRoleName = props.glueJobProps?.role!;
        } else {
            _glueJobRoleName = props.existingGlueJob.role;
        }

        this.kinesisStream.grantRead(Role.fromRoleArn(scope, 'GlueJobRole', _glueJobRoleName));
    }

    /**
     * This is a helper method to create the Role required for the Glue Job. If a role is already created then this
     * method is not required to be called.
     *
     * @param scope - The AWS Construct under which the role is to be created
     */
    public static createGlueJobRole(scope: Construct): Role {
        const _jobRole: Role = new Role(scope, 'JobRole', {
            assumedBy: new ServicePrincipal('glue.amazonaws.com'),
            description: 'Service role that Glue custom ETL jobs will assume for exeuction',
        });

        return _jobRole;
    }

    /**
     * This is a helper method to creates @CfnJob.JobCommandProperty for CfnJob. Based on the input parameters provided,
     * it will also create the S3 bucket for the ETL script location and grant 'read' access to 'glue.amazonaws.com'
     * Service Principal so that the script inside the bucket can be accessed as by AWS Glueglobal.fetch = require('node-fetch');
     *
     * Also this method does not set lifecycle policies on S3 buckets created unless they are explicitly set in the bucket
     * props
     *
     * @param scope - The AWS Construct under the underlying construct should be created
     * @param _jobID - The identifier/ name of the ETL Job
     * @param pythonVersion - The values as for Glue Documentation are '2' and '3'. There is no validation in the
     * method to check for these values to be forward compatible with Glue API changes. If valid values are not provided
     * the deployment would fail.
     * @param existingScriptLocation - If an S3 bucket location for the script exists, set this parameter. If the Bucket
     * is to be created, set the value as undefined. Setting this parameter will ignore @scriptLocationBucketProps as the
     * bucket already exists
     * @param scriptLocationBucketProps - Set this parameter only if the bucket is to be created. If the
     * @existingScriptLocation parameter is set, this parameter will be ignored. This parameter allows to set S3 Bucket
     * properts where the ETL script will be located
     */
    public static createGlueJobCommand(scope: Construct,
                                       _jobID: string, pythonVersion: string, existingScriptLocation?: Bucket,
                                       scriptLocationBucketProps?: BucketProps): CfnJob.JobCommandProperty {
        // create s3 bucket where script can be deployed
        let scriptLocation: [ Bucket, (Bucket | undefined)? ];
        if (existingScriptLocation === undefined) {
            scriptLocation = defaults.buildS3Bucket(scope, {
                bucketProps: scriptLocationBucketProps
            });
            // Remove the default LifecycleConfiguration for the Logging Bucket
            if (scriptLocationBucketProps?.lifecycleRules === undefined) {
                (scriptLocation[0]!.node.findChild('Resource') as CfnBucket).addPropertyDeletionOverride(
                                                                                        'LifecycleConfiguration.Rules');
            }
        } else {
            // since bucket location was provided in the props, logger bucket is not created
            scriptLocation = [ existingScriptLocation, undefined ];
        }

        scriptLocation[0].grantRead(new ServicePrincipal('glue.amazonaws.com'));

        const _jobCommand: CfnJob.JobCommandProperty = {
            name: _jobID,
            pythonVersion,
            scriptLocation: scriptLocation[0].s3UrlForObject()
        };

        return _jobCommand;
    }
}