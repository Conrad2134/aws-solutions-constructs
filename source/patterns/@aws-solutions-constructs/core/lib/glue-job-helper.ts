/**
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnDatabase, CfnJob, CfnJobProps, CfnSecurityConfiguration, CfnTable } from '@aws-cdk/aws-glue';
import { Effect, IRole, Policy, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import { Bucket, IBucket } from '@aws-cdk/aws-s3';
import { Aws, Construct, IResolvable } from '@aws-cdk/core';
import * as defaults from '../';
import { overrideProps } from './utils';

/**
 * Enumeration of data store types that could include S3, DynamoDB, DocumentDB, RDS or Redshift. Current
 * construct implementation only supports S3, but potential to add other output types in the future
 */
export enum SinkStoreType {
  S3 = 'S3'
}

/**
 * Interface to define potential outputs to allow the construct define additional output destinations for ETL
 * transformation
 */
export interface SinkDataStoreProps {
  /**
   * Sink data store type
   */
  readonly datastoreStype: SinkStoreType;
  /**
   * The output S3 location where the data should be written. The provided S3 bucket will be used to pass
   * the output location to the etl script as an argument to the AWS Glue job.
   *
   * If no location is provided the construct will
   * create a new S3 bucket location and pass it to the Glue job as arguments.
   *
   * The argument key is `output_path`. The value of the argument can be retrieve in the python script
   * as follows:
   *  getResolvedOptions(sys.argv, ["JOB_NAME", "output_path", <other arguments that are passed> ])
   *  output_path = args["output_path"]
   */
  readonly s3OutputBucket?: Bucket
}

export interface BuildGlueJobProps {
  /**
   * Glue ETL job properties. Do not pass the location of the script under the JobCommand. This
   * bucket location will be ignored and new location will be created. If a bucket location for the
   * ETL script exists, set it as the @scriptLocation parameter
   */
  readonly glueJobProps?: CfnJobProps | any
  /**
   * Existing instance of the S3 bucket object, if this is set then the script location is ignored.
   */
  readonly existingCfnJob?: CfnJob;

  readonly table: CfnTable;

  readonly database: CfnDatabase;

  readonly outputDataStore?: SinkDataStoreProps
}

export function buildGlueJob(scope: Construct, props: BuildGlueJobProps): [CfnJob, IRole] {

  if (!props.existingCfnJob) {
    if (props.glueJobProps) {
      return deployGlueJob(scope, props.glueJobProps, props.database!, props.table!, props.outputDataStore!);
    } else {
      throw Error('Either glueJobProps or existingCfnJob is required');
    }
  } else {
    return [props.existingCfnJob, Role.fromRoleArn(scope, 'ExistingRole', props.existingCfnJob.role)];
  }
}

export function deployGlueJob(scope: Construct, glueJobProps: CfnJobProps, database: CfnDatabase, table: CfnTable,
  outputDataStore: SinkDataStoreProps): [CfnJob, IRole] {

  let _glueSecurityConfigName: string;

  if (glueJobProps.securityConfiguration === undefined) {
    _glueSecurityConfigName = 'ETLJobSecurityConfig';
    const _glueKMSKey = `arn:${Aws.PARTITION}:kms:${Aws.REGION}:${Aws.ACCOUNT_ID}:alias/aws/glue`;

    new CfnSecurityConfiguration(scope, 'GlueSecurityConfig', {
      name: _glueSecurityConfigName,
      encryptionConfiguration: {
        jobBookmarksEncryption: {
          jobBookmarksEncryptionMode: 'CSE-KMS',
          kmsKeyArn: _glueKMSKey
        },
        s3Encryptions: [{
          s3EncryptionMode: 'SSE-S3'
        }]
      }
    });
  } else {
    _glueSecurityConfigName = glueJobProps.securityConfiguration;
  }

  const _glueJobPolicy = new Policy(scope, 'LogPolicy', {
    statements: [
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: [ 'logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents' ],
        resources: [ `arn:${Aws.PARTITION}:logs:${Aws.REGION}:${Aws.ACCOUNT_ID}:log-group:/aws-glue/*` ]
      })
    ]
  });

  let _jobRole: IRole;
  if (glueJobProps.role) {
    _jobRole = Role.fromRoleArn(scope, 'JobRole', glueJobProps.role);
  } else {
    _jobRole = defaults.createGlueJobRole(scope);
  }

  _glueJobPolicy.attachToRole(_jobRole);

  let _outputLocation: [ Bucket, Bucket? ];
  if (outputDataStore !== undefined && outputDataStore.s3OutputBucket !== undefined) {
    _outputLocation = [ outputDataStore.s3OutputBucket, undefined ];
  } else {
    _outputLocation = defaults.buildS3Bucket(scope, {});
  }

  const _jobArgumentsList = {
    "--enable-metrics" : true,
    "--enable-continuous-cloudwatch-log" : true,
    "--database_name": database.ref,
    "--table_name": table.ref,
    ...(outputDataStore && outputDataStore.datastoreStype === SinkStoreType.S3 &&
      { '--output_path' : `s3://${_outputLocation[0]}/output/` }),
    ...glueJobProps.defaultArguments
  };

  const _newGlueJobProps: CfnJobProps = overrideProps(defaults.DefaultGlueJobProps(_jobRole!, glueJobProps.command,
    _glueSecurityConfigName, _jobArgumentsList), glueJobProps);

  let _scriptLocation: string;
  if (isJobCommandProperty(_newGlueJobProps.command)) {
    if (_newGlueJobProps.command.scriptLocation) {
      _scriptLocation = _newGlueJobProps.command.scriptLocation;
    } else {
      throw Error('Script location has to be provided as an s3 Url location. Script location cannot be empty');
    }
  }

  const _scriptBucketLocation: IBucket = Bucket.fromBucketArn(scope, 'ScriptLocaiton', getS3ArnfromS3Url(_scriptLocation!));
  _scriptBucketLocation.grantRead(_jobRole);

  const _glueJob: CfnJob = new CfnJob(scope, 'KinesisETLJob', _newGlueJobProps);
  return [_glueJob, _jobRole];
}

/**
 * This is a helper method to create the Role required for the Glue Job. If a role is already created then this
 * method is not required to be called.
 *
 * @param scope - The AWS Construct under which the role is to be created
 */
export function createGlueJobRole(scope: Construct): Role {
  return new Role(scope, 'JobRole', {
    assumedBy: new ServicePrincipal('glue.amazonaws.com'),
    description: 'Service role that Glue custom ETL jobs will assume for exeuction',
  });
}

export function createGlueTable(scope: Construct, database: CfnDatabase, fieldSchema: CfnTable.ColumnProperty [], sourceType: string,
  parameters?: any): CfnTable {
  return defaults.DefaultGlueTable(scope, database, fieldSchema, sourceType, parameters);
}

export function createGlueDatabase(scope: Construct): CfnDatabase {
  return defaults.DefaultGlueDatabase(scope);
}

function getS3ArnfromS3Url(s3Url: string): string {
  const splitString: string = s3Url.slice(5);
  return `arn:${Aws.PARTITION}:s3:::${splitString}`;
}

function isJobCommandProperty(command: CfnJob.JobCommandProperty | IResolvable): command is CfnJob.JobCommandProperty {
  if ((command as CfnJob.JobCommandProperty).name ||
    (command as CfnJob.JobCommandProperty).pythonVersion ||
    (command as CfnJob.JobCommandProperty).scriptLocation) {

    return true;
  } else {
    return false;
  }
}