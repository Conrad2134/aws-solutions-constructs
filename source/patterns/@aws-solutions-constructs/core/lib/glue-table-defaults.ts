/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnDatabase, CfnTable } from "@aws-cdk/aws-glue";
import { Aws, Construct } from "@aws-cdk/core";

/**
 *
 * @param scope
 * @param database
 * @param fieldSchema
 * @param sourceType
 * @param parameters - Key value pairs of parmaeters. If source type is 'Kinesis', pass Kinesis Data Stream name with key
 *  as 'STREAM_NAME'. Example: {STREAM_NAME: 'KinesisStreamConstrct.streamName'}
 */
export function DefaultGlueTableProps(scope: Construct, database: CfnDatabase, fieldSchema: CfnTable.ColumnProperty [],
                                      sourceType: string, parameters?: any): CfnTable {
  let _glueTable: CfnTable;
  if (sourceType === 'Kinesis') {
    const kinesisStreamName = parameters.STREAM_NAME;
    _glueTable = new CfnTable(scope, 'GlueTable', {
      catalogId: database.catalogId,
      databaseName: database.ref,
      tableInput: {
        storageDescriptor: {
          columns: fieldSchema,
          location: kinesisStreamName,
          inputFormat: "org.apache.hadoop.mapred.TextInputFormat",
          outputFormat: "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
          compressed: false,
          numberOfBuckets: -1,
          serdeInfo: {
            serializationLibrary: "org.openx.data.jsonserde.JsonSerDe",
            parameters: {
              paths: '.'
            }
          },
          parameters: {
            endpointUrl: `https://kinesis.${Aws.REGION}.amazonaws.com`,
            streamName: kinesisStreamName,
            typeOfData: "kinesis"
          }
        },
        tableType: 'EXTERNAL_TABLE',
        parameters: {
          classication: 'json'
        }
      }
    });
  } else {
    throw Error('Source Type not Supported. Valid Source Type not provided');
  }

  return _glueTable!;
}