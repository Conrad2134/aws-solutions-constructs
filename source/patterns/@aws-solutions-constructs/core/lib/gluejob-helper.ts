/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { CfnJob, CfnJobProps, CfnSecurityConfiguration } from '@aws-cdk/aws-glue';
import { Effect, Policy, PolicyStatement, Role } from '@aws-cdk/aws-iam';
import { Aws, Construct } from '@aws-cdk/core';
import * as defaults from '../';
import { overrideProps } from './utils';

export interface BuildGlueJobProps {
  /**
   * Glue ETL job properties. Do not pass the location of the script under the JobCommand. This
   * bucket location will be ignored and new location will be created. If a bucket location for the
   * ETL script exists, set it as the @scriptLocation parameter
   */
  readonly glueJobProps?: CfnJobProps
  /**
   * Existing instance of the S3 bucket object, if this is set then the script location is ignored.
   */
  readonly existingCfnJob?: CfnJob;
}

export function buildGlueJob(scope: Construct, props: BuildGlueJobProps): CfnJob {

  if (!props.existingCfnJob) {
    if (props.glueJobProps) {
      return deployGlueJob(scope, props.glueJobProps);
    } else {
      throw Error('Either glueJobProps or existingCfnJob is required');
    }
  } else {
    // return properties are already supplied then bucket is not created and hence returns undefined
    return props.existingCfnJob;
  }
}

export function deployGlueJob(scope: Construct, glueJobProps: CfnJobProps): CfnJob {
  const _jobID = 'ETLJob';

  let _glueSecurityConfigName: string;

  if (glueJobProps.securityConfiguration === undefined) {
    _glueSecurityConfigName = 'ETLJobSecurityConfig';
    const _glueKMSKey = `arn:${Aws.PARTITION}:kms:${Aws.REGION}:${Aws.ACCOUNT_ID}:alias/aws/glue`;

    new CfnSecurityConfiguration(scope, 'GlueSecurityConfig', {
      name: _glueSecurityConfigName,
      encryptionConfiguration: {
        cloudWatchEncryption: {
          cloudWatchEncryptionMode: 'SSE-KMS',
          kmsKeyArn: _glueKMSKey
        },
        jobBookmarksEncryption: {
          jobBookmarksEncryptionMode: 'CSE-KMS',
          kmsKeyArn: _glueKMSKey
        },
        s3Encryptions: [{
          s3EncryptionMode: 'SSE-S3'
        }]
      }
    });
  } else {
    _glueSecurityConfigName = glueJobProps.securityConfiguration;
  }

  const _glueJobPolicy = new Policy(scope, 'LogPolicy', {
    statements: [
      new PolicyStatement({
        effect: Effect.ALLOW,
        actions: [ 'logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents' ],
        resources: [ `arn:${Aws.PARTITION}:logs:${Aws.REGION}:${Aws.ACCOUNT_ID}:log-group/aws-glue/*` ]
      })
    ]
  });

  _glueJobPolicy.attachToRole(Role.fromRoleArn(scope, 'GlueJobRole', glueJobProps.role));

  const _glueJobProps: CfnJobProps = overrideProps(defaults.DefaultGlueJobProps(glueJobProps.role, glueJobProps.command,
                                                                              _glueSecurityConfigName), glueJobProps);

  const _glueJob: CfnJob = new CfnJob(scope, _jobID, _glueJobProps);
  return _glueJob;
}