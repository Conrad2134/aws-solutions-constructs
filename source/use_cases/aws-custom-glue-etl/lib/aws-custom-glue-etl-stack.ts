/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

 import { StreamEncryption } from '@aws-cdk/aws-kinesis';
import * as cdk from '@aws-cdk/core';
import { Aws, CfnOutput } from '@aws-cdk/core';
import { KinesisStreamGlueJob } from '@aws-solutions-constructs/aws-kinesisstreams-gluejob';
import * as defaults from '@aws-solutions-constructs/core';

export class AwsCustomGlueEtlStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const _glueJobRole = KinesisStreamGlueJob.createGlueJobRole(this);

    const _outputBucket = defaults.buildS3Bucket(this, {
      bucketProps: defaults.DefaultS3Props()
    });

    _outputBucket[0].grantReadWrite(_glueJobRole);

    const _jobCommand = KinesisStreamGlueJob.createGlueJobCommand(this, 'glueetl', '3', undefined, `${__dirname}/../etl/transform.py`);

    const _customEtlJob = new KinesisStreamGlueJob(this, 'CustomETL', {
      kinesisStreamProps: {
       encryption: StreamEncryption.UNENCRYPTED
      },
      glueJobProps: {
        command: _jobCommand[0],
        role: _glueJobRole.roleArn,
        defaultArguments: {
          "--job-bookmark-option": "job-bookmark-enable",
          "--enable-metrics" : true,
          "--enable-continuous-cloudwatch-log" : true,
          "--enable-glue-datacatalog": true,
          "--output-path": `s3://${_outputBucket[0].bucketName}/output/`,
          "--aws-region": Aws.REGION
        }
      }
    });

    new CfnOutput(this, 'KinesisStreamName', {
      value: _customEtlJob.kinesisStream.streamName
    });

    new CfnOutput(this, 'ScripLocation', {
      value: _jobCommand[1]!.s3ObjectUrl
    });

    new CfnOutput(this, 'OutputBucket', {
      value: _outputBucket[0].bucketArn
    });
  }
}