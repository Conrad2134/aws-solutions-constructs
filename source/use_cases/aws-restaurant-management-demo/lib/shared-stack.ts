/**
 *  Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import * as cdk from '@aws-cdk/core';
import * as dynamodb from '@aws-cdk/aws-dynamodb';

// Stack
export class SharedStack extends cdk.Stack {

  // Public variables
  public readonly database: dynamodb.Table;

  // Constructor
  constructor(scope: cdk.Construct, id: string, props: cdk.StackProps) {
    super(scope, id, props);

    // Database for storing orders
    this.database = new dynamodb.Table(this, 'database', {
      partitionKey: {
        name: 'id',
        type: dynamodb.AttributeType.STRING
      }
    });
  }
}