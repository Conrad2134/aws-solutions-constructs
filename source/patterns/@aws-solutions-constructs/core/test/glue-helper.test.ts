/**
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

// Imports
import { SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import { CfnJobProps } from '@aws-cdk/aws-glue';
import { Stack } from "@aws-cdk/core";
import { DefaultGlueJobProps } from '../../core/lib/gluejob-defaults';
import { buildGlueJob } from '../../core/lib/gluejob-helper';

// --------------------------------------------------------------
// Test minimal deployment with no properties
// --------------------------------------------------------------
test('Test minimal deployment with no properties', () => {
    // Stack
    const stack = new Stack();

    const _jobID = 'testETLJob';

    const cfnJobProps: CfnJobProps = DefaultGlueJobProps('jobRole', {
        name: _jobID,
        pythonVersion: '3',
        scriptLocation: 's3://fakelocation/script'
    }, _jobID);

    buildGlueJob(stack, { glueJobProps: cfnJobProps });
    // Assertion 1
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    // Assertion 2
    // expect(stack).toHaveResourceLike('AWS::Glue::Job', {
    //     Properties: {
    //         Command: {
    //             Name : 'testETLJob',
    //             PythonVersion : '3',
    //             ScriptLocation : 's3://fakelocation/script'
    //         },
    //         Role: "jobRole",
    //         SecurityConfiguration: 'testETLJob'
    //     },
    //     Type: "AWS::Glue::Job",
    // });
});