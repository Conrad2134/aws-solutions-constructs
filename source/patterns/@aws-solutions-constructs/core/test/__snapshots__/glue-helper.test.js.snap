// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create a Glue Job outside the construct 1`] = `
Object {
  "Resources": Object {
    "ExistingJob": Object {
      "Properties": Object {
        "AllocatedCapacity": 2,
        "Command": Object {
          "Name": "existingJob",
          "PythonVersion": "2",
          "ScriptLocation": "s3://existingFakeLocation/existingScript",
        },
        "GlueVersion": "1",
        "MaxCapacity": 4,
        "NumberOfWorkers": 2,
        "Role": Object {
          "Fn::GetAtt": Array [
            "ExistingJobRole8F750976",
            "Arn",
          ],
        },
        "WorkerType": "Standard",
      },
      "Type": "AWS::Glue::Job",
    },
    "ExistingJobRole8F750976": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazon.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Existing role",
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`Test custom deployment properties 1`] = `
Object {
  "Resources": Object {
    "ETLJob": Object {
      "Properties": Object {
        "AllocatedCapacity": 2,
        "Command": Object {
          "Name": "testETLJob",
          "PythonVersion": "3",
          "ScriptLocation": "s3://existingFakeLocation/existingScript",
        },
        "GlueVersion": "1",
        "MaxCapacity": 4,
        "NumberOfWorkers": 2,
        "Role": Object {
          "Fn::GetAtt": Array [
            "ExistingJobRole8F750976",
            "Arn",
          ],
        },
        "SecurityConfiguration": "ETLJobSecurityConfig",
        "WorkerType": "Standard",
      },
      "Type": "AWS::Glue::Job",
    },
    "ExistingJobRole8F750976": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazon.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Existing role",
      },
      "Type": "AWS::IAM::Role",
    },
    "GlueSecurityConfig": Object {
      "Properties": Object {
        "EncryptionConfiguration": Object {
          "CloudWatchEncryption": Object {
            "CloudWatchEncryptionMode": "SSE-KMS",
            "KmsKeyArn": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":kms:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":alias/aws/glue",
                ],
              ],
            },
          },
          "JobBookmarksEncryption": Object {
            "JobBookmarksEncryptionMode": "CSE-KMS",
            "KmsKeyArn": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "arn:",
                  Object {
                    "Ref": "AWS::Partition",
                  },
                  ":kms:",
                  Object {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  Object {
                    "Ref": "AWS::AccountId",
                  },
                  ":alias/aws/glue",
                ],
              ],
            },
          },
          "S3Encryptions": Array [
            Object {
              "S3EncryptionMode": "SSE-S3",
            },
          ],
        },
        "Name": "ETLJobSecurityConfig",
      },
      "Type": "AWS::Glue::SecurityConfiguration",
    },
    "LogPolicy9292E033": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group/aws-glue/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogPolicy9292E033",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "ExistingJobRole8F750976",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`Test minimal deployment with no properties 1`] = `
Object {
  "Resources": Object {
    "CustomETLJobRole90A83A66": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ETLJob": Object {
      "Properties": Object {
        "Command": Object {
          "Name": "testETLJob",
          "PythonVersion": "3",
          "ScriptLocation": "s3://fakelocation/script",
        },
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomETLJobRole90A83A66",
            "Arn",
          ],
        },
        "SecurityConfiguration": "testETLJob",
      },
      "Type": "AWS::Glue::Job",
    },
    "LogPolicy9292E033": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group/aws-glue/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogPolicy9292E033",
        "Roles": Array [
          Object {
            "Fn::Select": Array [
              1,
              Object {
                "Fn::Split": Array [
                  "/",
                  Object {
                    "Fn::Select": Array [
                      5,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::GetAtt": Array [
                              "CustomETLJobRole90A83A66",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
